dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/AbstractCondition.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.tei.query;
+
+import static org.hisp.dhis.analytics.tei.query.QueryContextConstants.ENR_ALIAS;
+import static org.hisp.dhis.analytics.tei.query.QueryContextConstants.EVT_ALIAS;
+import static org.hisp.dhis.analytics.tei.query.QueryContextConstants.TEI_ALIAS;
+
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.analytics.common.dimension.DimensionIdentifier;
+import org.hisp.dhis.analytics.common.dimension.DimensionParam;
+import org.hisp.dhis.analytics.common.query.*;
+import org.hisp.dhis.analytics.tei.query.context.QueryContext;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramStage;
+
+@RequiredArgsConstructor
+public abstract class AbstractCondition extends BaseRenderable
+{
+
+    private final DimensionIdentifier<Program, ProgramStage, DimensionParam> dimensionIdentifier;
+
+    private final QueryContext queryContext;
+
+    @Override
+    public String render()
+    {
+        if ( dimensionIdentifier.isEnrollmentDimension() )
+        {
+            return enrollmentConditionInternal();
+        }
+        if ( dimensionIdentifier.isEventDimension() )
+        {
+            return conditionForEvent();
+        }
+        return conditionForTei();
+    }
+
+    private String conditionForTei()
+    {
+        return getTeiCondition().render();
+    }
+
+    protected Renderable getTeiCondition()
+    {
+        // Default implementation, subclasses will override where needed
+        throw new IllegalStateException( "Not supported for TEI dimension" );
+    };
+
+    private String enrollmentConditionInternal()
+    {
+        String programUid = dimensionIdentifier.getProgram().getElement().getUid();
+        return "exists (select 1 from " + getEnrollmentSubQuery( programUid ) + " where "
+            + getEnrollmentCondition().render() + ")";
+    }
+
+    protected Renderable getEnrollmentCondition()
+    {
+        // Default implementation, subclasses will override where needed
+        throw new IllegalStateException( "Not supported for enrollment dimension" );
+    }
+
+    private String getEnrollmentSubQuery( String programUid )
+    {
+        return " (SELECT *" +
```

Comment:
SQL statements should be all lowercase per our coding standards.
The same applies to all other SQL statements in this PR.